libtock/alarm_timer.c

```diff
@@ -123,45 +243,89 @@ void alarm_cancel(alarm_t* alarm) {
 
   alarm->prev = NULL;
   alarm->next = NULL;
+}
 
+// The intermediate callback that handles overflows of alarm. This is used by
+// `timer_in` to handle cases where `alarm_at` would normally be unable to
+// support the whole timer length. `alarm_at` can only keep track of up to 2^32
+// ticks.
+static void overflow_callback(int                          last_timer_fire_time,
+                              __attribute__ ((unused)) int unused1,
+                              __attribute__ ((unused)) int unused2,
+                              void*                        overflow_ud) {
+  tock_timer_t* tock_timer = (tock_timer_t*)overflow_ud;
+
+  if (tock_timer->overflows_left == 0) {
+    // no overflows left, schedule last alarm with original callback
+    alarm_at(
+      last_timer_fire_time,
+      tock_timer->remaining_ticks,
+      tock_timer->callback,
+      tock_timer->user_data,
+      &(tock_timer->alarm)
+      );
+  } else {
+    // schedule next intermediate alarm that will overflow
+    tock_timer->overflows_left--;
+
+    alarm_at(
+      last_timer_fire_time,
+      MAX_TICKS,
+      (subscribe_upcall*) overflow_callback,
+      (void*) tock_timer,
+      &(tock_timer->alarm)
+      );
+  }
 }
 
 // Timer implementation
-
-int timer_in(uint32_t ms, subscribe_upcall cb, void* ud, tock_timer_t *timer) {
-  uint32_t frequency;
-  alarm_internal_frequency(&frequency);
-  uint32_t interval = (ms / 1000) * frequency + (ms % 1000) * (frequency / 1000);
+int timer_in(uint32_t dt_ms, subscribe_upcall cb, void* ud, tock_timer_t *tock_timer) {
```

Comment:
Why did the argument name need to change? Now it doesn't match the header file.

Reply:
Resolved here: https://github.com/tock/libtock-c/pull/395/commits/6277a8924e0d4f3b0942deb1e032ce9314784136

`timer_in` arguments are consistent with the header.